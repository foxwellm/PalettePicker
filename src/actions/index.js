export const setCurrentPalette = (palette) => ({
  type: 'SET_CURRENT_PALETTE',
  palette
});

export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const addProject = (project) => ({
  type: 'ADD_PROJECT',
  project
});

export const removeProject = (projectId) => ({
  type: 'REMOVE_PROJECT',
  projectId
});

export const setPalettes = (palettes) => ({
  type: 'SET_PALETTES',
  palettes
});

export const addPalette = (palette) => ({
  type: 'ADD_PALETTE',
  palette
});

export const removePalette = (paletteId) => ({
  type: 'REMOVE_PALETTE',
  paletteId
});

export const setSearchProjects = (search) => ({
  type: 'SEARCH_PROJECTS',
  search
});

export const setSearching = (bool) => ({
  type: 'SEARCHING_PROJECTS',
  bool
});

export const setLoading = (loading) => ({
  type: 'SET_LOADING',
  loading
});

export const setError = (error) => ({
  type: 'SET_ERROR',
  error
});