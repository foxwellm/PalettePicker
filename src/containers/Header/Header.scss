#title {
  width: 300px;
  display: block;
  font-size: 60px;
  font-weight: 800;
  color: black;
  text-decoration: none;
  margin: 20px auto;
  font-size: 45px;
}

.search-bar {
  position: absolute;
  top: 20px;
  right: 30px;
}