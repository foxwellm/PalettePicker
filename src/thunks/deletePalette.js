import { setLoading, setError, removePalette } from '../actions';
import { fetchData } from '../utils/api';

export const deletePalette = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      await fetchData(`/api/v1/palettes/${id}`, 'DELETE');
      dispatch(removePalette(id));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}