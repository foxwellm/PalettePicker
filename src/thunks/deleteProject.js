import { setLoading, setError, removeProject } from '../actions';
import { fetchData } from '../utils/api';

export const deleteProject = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      await fetchData(`/api/v1/projects/${id}`, 'DELETE');
      dispatch(removeProject(id));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}