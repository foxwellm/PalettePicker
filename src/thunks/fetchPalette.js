import { setLoading, setError, setPalettes } from '../actions';
import { fetchData } from '../utils/api';

export const fetchPalette = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const palette = await fetchData(`/api/v1/palettes/${id}`, 'GET');
      dispatch(setPalettes([palette]));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  };
};