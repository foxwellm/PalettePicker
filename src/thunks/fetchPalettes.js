import { setError } from '../actions';
import { fetchData } from '../utils/api';

export const fetchPalettes = (allProjects) => {
  return (dispatch) => {
    const promisedPalettes = allProjects.map(async palette => {
      try {
        let allPalettes = await fetchData(`/api/v1/projects/${palette.id}/palettes`, 'GET');
        allPalettes.sort((a, b) => {
          if (a.updated_at > b.updated_at) return -1;
          if (b.updated_at < a.updated_at) return 1;
          return 0;
        });
        return allPalettes
      } catch (error) {
        dispatch(setError(error.message));
      }
    })
    return Promise.all(promisedPalettes);
  }
}