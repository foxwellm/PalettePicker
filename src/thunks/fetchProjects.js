import { setLoading, setError, setProjects } from '../actions';
