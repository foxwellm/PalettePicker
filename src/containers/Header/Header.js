import React, { Component } from 'react';
import { connect } from 'react-redux';
import { searchProjects } from '../../thunks/searchProjects';
import { setSearching } from '../../actions';

export class Header extends Component {
  constructor(props) {
    super(props);
    this.state = {
      search: ''
    };
  };

  handleChange = (e) => {
    const { value } = e.target;
    const { setSearching, searchProjects } = this.props;
    const { search } = this.state;

    this.setState({ search: value },
      () => {
        if (value.length) {
          setSearching(true)
          searchProjects(search)
        } else {
          setSearching(false)
        }
      });
  };

  render() {
    const { search } = this.state;
    return (
      <div className='header-div'>
        <h1>Palette Picker</h1>
        <label className='search-bar'>Search Projects:
        <input onChange={this.handleChange} value={search} placeholder='Type project/palette name'></input>
        </label>
      </div>
    );
  };
};

export const mapStateToProps = (state) => ({
  searching: state.searching,
});

export const mapDispatchToProps = (dispatch) => ({
  searchProjects: (search) => dispatch(searchProjects(search)),
  setSearching: (bool) => dispatch(setSearching(bool)),
});


export default connect(mapStateToProps, mapDispatchToProps)(Header);