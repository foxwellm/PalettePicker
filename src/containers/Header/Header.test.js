import React from 'react';
import { Header, mapDispatchToProps } from './Header';
import { shallow } from 'enzyme';
import { fetchProjects } from '../../thunks/fetchProjects';
import { setError, setSearching } from '../../actions';
jest.mock('../../thunks/fetchProjects');

describe('Header', () => {
  let wrapper;
  let fetchProjectsMock;
  let setErrorMock;
  let setSearchingMock;
  let mockEvent;

  beforeEach(() => {
    fetchProjectsMock = jest.fn();
    setErrorMock = jest.fn();
    setSearchingMock = jest.fn();
    mockEvent = { target: { value: 'test'} };

    wrapper = shallow(
      <Header
        setError={setErrorMock}
        fetchProjects={fetchProjectsMock}
        setSearching={setSearchingMock}
      />
    )
  });

  it('should match the correct snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default state', () => {
    expect(wrapper.state()).toEqual({ search: ''});
  });

  describe('handleChange', () => {
    it('should call setError with the correct params', () => {
      wrapper.instance().handleChange(mockEvent);
      expect(setErrorMock).toHaveBeenCalledWith('');
    });

    it('should update state with the correct value', () => {
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('search')).toEqual('test');
    });

    it('should call handleSearch', () => {
      wrapper.instance().handleSearch = jest.fn();
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.instance().handleSearch).toHaveBeenCalled();
    });
  });

  describe('handleSearch', () => {
    it('should call setSearching with false if no search', () => {
      wrapper.instance().handleSearch(mockEvent);
      expect(setSearchingMock).toHaveBeenCalledWith(false);
    });

    it('should call fetchProjects with no params if no search', () => {
      wrapper.instance().handleSearch(mockEvent);
      expect(fetchProjectsMock).toHaveBeenCalledWith();
    });

    it('should call setSearching with true if search', () => {
      wrapper.setState({ search: 'test'});
      wrapper.instance().handleSearch(mockEvent);
      expect(setSearchingMock).toHaveBeenCalledWith(true);
    });

    it('should call fetchProjects with the correct params if search', () => {
      wrapper.setState({ search: 'test' });
      wrapper.instance().handleSearch(mockEvent);
      expect(fetchProjectsMock).toHaveBeenCalledWith('test');
    });

    it('should setState the correct value', () => {
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('search')).toEqual('test');
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch when setError is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setError();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setError();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch when fetchProjects is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchProjects();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchProjects();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch when setSearching is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setSearching();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setSearching();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});