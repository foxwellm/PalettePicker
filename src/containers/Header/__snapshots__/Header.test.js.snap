// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match the correct snapshot 1`] = `
<div
  className="header-div"
>
  <h1>
    Palette Picker
  </h1>
  <label
    className="search-bar"
  >
    Search Projects:
    <input
      onChange={[Function]}
      placeholder="Type project/palette name"
      value=""
    />
  </label>
</div>
`;
